### PR TITLE
feat: add PolynomialLR scheduler

### LINKED ISSUE #79511: Suggest for a polynominal lr_scheduler
### 🚀 The feature, motivation and pitch

A polynominal lr_schduler is useful in multiple task.

### Alternatives

_No response_

### Additional context

https://github.com/cmpark0126/pytorch-polynomial-lr-decay
https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/schedules/PolynomialDecay

Here is a possible useful repo for this issue.

cc @vincentqb @jbschlosser @albanD

### PR DESCRIPTION
**Pull request by federicopozzi33**
### Description

Add PolynomialLR scheduler. 

### Issue
Closes #79511.

### Testing
I added tests for PolynomialLR.

### CODE REVIEW

**Comment by federicopozzi33** [comment]
@datumbox

First draft.

Tests can be improved (combined versions are missing, e.g. `PolynomialLR` + `StepLR`).

I tried to generate the `.pyi` for `optimizer.py` but the result I obtained is quite different from the existing version. This is what I did, using `mypy`:

```bash
stubgen torch/optim/lr_scheduler.py
```

Generated file:
```python
class PolynomialLR(_LRScheduler):
    total_iters: Incomplete
    min_lrs: Incomplete
    power: Incomplete
    def __init__(self, optimizer, total_iters: int = ..., min_lr: float = ..., power: float = ..., last_epoch: int = ..., verbose: bool = ...) -> None: ...
    def get_lr(self): ...

class LinearLR(_LRScheduler):
    start_factor: Incomplete
    end_factor: Incomplete
    total_iters: Incomplete
    def __init__(self, optimizer, start_factor=..., end_factor: float = ..., total_iters: int = ..., last_epoch: int = ..., verbose: bool = ...) -> None: ...
    def get_lr(self): ...
```

Existing `optimizer.pyi`:

```python
class LinearLR(_LRScheduler):
    start_factor: float = ...
    end_factor: float = ...
    total_iters: int = ...
    def __init__(self, optimizer: Optimizer, start_factor: float=..., end_factor: float= ..., total_iters: int= ..., last_epoch: int= ..., verbose: bool = ...) -> None: ...
```

File: test/test_optim.py
Comment (datumbox) [inline test/test_optim.py:1352]: Perhaps it's worth adding a test that passes a list to the `min_lr` parameter. You will need multiple groups of parameters for that. See a similar Scheduler that does this at:
https://github.com/pytorch/pytorch/blob/14dba3d70208f5c6e4f1ea857c25d32a1d9d06d1/test/test_optim.py#L1497

File: test/test_optim.py
Comment (datumbox) [inline test/test_optim.py:1360]: Could this scheduler and especially its `min_lr` feature have an interaction with some of the existing schedulers? In particular I'm thinking what happens if it's combined with another scheduler using `ChainedScheduler`. It would be good to add a test that confirms the scheduler produces the expected result in such a scenario.

BTW a combo with `StepLR` is not needed here. The reason why the example PR that I sent you was doing it was because it was introducing a Warmup scheduler that could have interactions with it.

File: torch/optim/lr_scheduler.py
Comment (datumbox) [inline torch/optim/lr_scheduler.py:710]:
```suggestion
        total_iters (int): The number of steps that the scheduler decays the learning rate.
            Default: 5.
```

File: torch/optim/lr_scheduler.py
Comment (datumbox) [inline torch/optim/lr_scheduler.py:714]:
```suggestion
        power (int): The power of the polynomial. Default: ``1.0``.
```

File: torch/optim/lr_scheduler.py
Comment (datumbox) [inline torch/optim/lr_scheduler.py:724]: Are sure these values are right? I believe for the proposed configuration it would be more like:
```
0 0.001
1 0.00075
2 0.0005
3 0.00025
```

File: torch/optim/lr_scheduler.py
Comment (datumbox) [inline torch/optim/lr_scheduler.py:703]: Make sure you add this on the `__all__` at the beginning of the file.

**Review by datumbox — COMMENTED** [review]
@federicopozzi33 Thanks a lot for the PR. As discussed at https://github.com/pytorch/vision/issues/4438 this is definitely something that TorchVision can use.

I've added a few comments but overall it looks good. Please let me know what you think.

> Documentation should be quite ok, but building documentation phase is missing

I had to manually approve running your tests because that was your first PR. It's running the jobs now.

> I confused torchvision issue with torch one, so the issue number in the branch name is wrong. Should I delete the branch?

I don't think PyTorch core has such a strict process for the naming conventions on your repo/branch. If renaming is needed we can do after the reviews.

> I tried to generate the .pyi for optimizer.py but the result I obtained is quite different from the existing version. 

I'm not sure about this one, let's leave it last to fix. Once the PR is ready, I'll ping some Core devs who could clarify these for us. :)

File: test/test_optim.py
Comment (federicopozzi33) [reply test/test_optim.py:1360]: For some reasons, `PolynomialLR` combined with another scheduler (e.g. `ConstantLR`) using `ChainedScheduler` + `min_lr` does not produce the expected values. I'll investigate more.

In the meantime, I added a test without `min_lr`.

File: torch/optim/lr_scheduler.py
Comment (federicopozzi33) [reply torch/optim/lr_scheduler.py:714]: Why the default values here is ``1.0``, while `total_iters` is just `5`?

File: torch/optim/lr_scheduler.py
Comment (federicopozzi33) [reply torch/optim/lr_scheduler.py:724]: I checked the values. You're right. Corrected.

File: test/test_optim.py
Comment (datumbox) [reply test/test_optim.py:1360]: Sounds good. Let me know when you have more info.

BTW I would be fine to drop the `min_lr` feature here. I maintained it on my draft because it was available on one of the references but overall not many schedulers implement it. Also I believe that the functionality of `min_lr` can be implemented more effectively by using the `ChainedScheduler` to effectively set it to a minimum fixed value.

File: torch/optim/lr_scheduler.py
Comment (datumbox) [reply torch/optim/lr_scheduler.py]: Thanks for pointing out. It seems that PyTorch Core doesn't use backticks for numerical values. Feel free to skip them.

File: test/test_optim.py
Comment (federicopozzi33) [reply test/test_optim.py:1360]: Seems pretty reasonable. I looked at some other similar schedulers like `ExponentialLR` or `ConstantLR` and they don't have that feature.

However, the `total_steps` parameter is now more like a "Number of steps needed to reach lr = 0". That's because the polynomial decay function decreases the initial learning rate to 0 in `total_steps`.

What do you think?

Comment (datumbox) [reply test/test_optim.py:1360]: Your remark about the new interpretation of `total_steps` if you remove `min_lr` is accurate. I would add that the current description on doc would be more accurate in this case.

But please note that you shouldn't let my original draft implementation bias the way you think the scheduler should be structured. The reason I provided it was to highlight specific parts of the API of the `_LRScheduler` class and in particular how to write `get_lr()` and `_get_closed_form_lr()`. If you think there is a better and more natural way to structure it, I would love to see it.

Comment (federicopozzi33) [reply test/test_optim.py:1360]: I agree with your implementation, that's why I didn't change it. The only doubt was about how to write the `get_lr` method, in particular how to compute the next decayed LR starting from the previous one, i.e. the current values in the optimizer (and not from the initial one), but you replied to me.

The `_get_closed_form_lr` implementation is pretty standard, I double-checked with another [source](https://www.tensorflow.org/api_docs/python/tf/compat/v1/train/polynomial_decay) and both use the same formula.

What is the reason behind two different implementations? Shouldn't the `get_lr` method compute the decayed learning rate always from the initial optimizer learning rate as in the `get_closed_form_lr`? It would make the scheduler simpler to understand and less prone to error (the polynomial function would be only one instead of two different versions like now).

Comment (datumbox) [reply test/test_optim.py:1360]: > IMO, it's not clear that the end learning rate is 0.

We can review the doc descriptions before you are done to ensure this is clear.

> What is the reason behind two different implementations?

Unfortunately I don't have the historical context on why this API is implemented this way. I don't regularly work on this code-base so there are lots of things I'm not familiar with. Perhaps one of the PyTorch Core devs can provide more info here.

Comment (federicopozzi33) [reply test/test_optim.py:1360]: Ok. 

I will remove the `min_lr` parameter and fix the latest comments by tomorrow. Then, I think it's ready for one more round of reviews.

File: torch/optim/lr_scheduler.py
Comment (federicopozzi33) [reply torch/optim/lr_scheduler.py]: Done.

File: test/test_optim.py
Comment (federicopozzi33) [reply test/test_optim.py:1360]: > Ok.
> 
> I will remove the `min_lr` parameter and fix the latest comments by tomorrow. Then, I think it's ready for one more round of reviews.

Done.

File: torch/optim/lr_scheduler.py
Comment (datumbox) [inline torch/optim/lr_scheduler.py:748]: Nit: We no longer need enumerate:
```suggestion
            for base_lr in self.base_lrs
```

**Review by datumbox — COMMENTED** [review]
I think overall it looks good. I've added only one minor nit comment below.

The question around the generation of `optimizer.pyi` remains, but we can ask core devs how to handle it. I've restarted the unit-tests so that we can get a signal from the CI. If all tests pass, from my perspective, we should mark the PR as non draft and ask for the review of Alban/Joel.

**Comment by federicopozzi33** [comment]
> I think overall it looks good. I've added only one minor nit comment below.
> 
> The question around the generation of `optimizer.pyi` remains, but we can ask core devs how to handle it. I've restarted the unit-tests so that we can get a signal from the CI. If all tests pass, from my perspective, we should mark the PR as non draft and ask for the review of Alban/Joel.

Some tests are still failing, but it seems that they are not related to my changes... can you check you too, please?

**Comment by datumbox** [comment]
LGTM! The [failing](https://github.com/pytorch/pytorch/runs/7722122835?check_suite_focus=true) tests are unrelated. 

I think now we should be good to ask for the feedback of @albanD and @jbschlosser. Please let us know your thoughts and your recommendation for the generation of `optimizer.pyi`.

File: torch/optim/lr_scheduler.py
Comment (albanD) [inline torch/optim/lr_scheduler.py:716]: Should this be `0.001` ?

**Review by albanD — COMMENTED** [review]
Small typo in the doc. but LGTM otherwise.

Thanks @datumbox for doing the early review!

**Comment by albanD** [comment]
The `.pyi` files are written by hand if they are checked into the git repo. That's why it doesn't match what mypy would generate.

File: torch/optim/lr_scheduler.py
Comment (datumbox) [reply torch/optim/lr_scheduler.py:716]: Ah that's a remnant from https://github.com/pytorch/pytorch/pull/82769#discussion_r937552195

We should change it to `0.00010`. cc @federicopozzi33

Comment (federicopozzi33) [reply torch/optim/lr_scheduler.py]: > Ah that's a remnant from [#82769 (comment)](https://github.com/pytorch/pytorch/pull/82769#discussion_r937552195)
> 
> We should change the lr on the sentence to `0.0001`. cc @federicopozzi33

The `lr` on the sentence (_Assuming optimizer uses lr = 0.001 for all groups_) is right. The `lr` at the first epoch was wrong (before `0.0001`, now `0.001`).

**Review by datumbox — APPROVED** [review]
LGTM, thanks @federicopozzi33. 

I'll leave it to @albanD and @jbschlosser to approve and merge. Note that I spoke with them and they told me it's quite hectic for them this week, so we might need to be a bit patient. Don't worry I'll follow up with them regularly to get this over the line.

**Pull request closed without merging**
